pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -304,6 +300,12 @@ pub struct FreezeLayers {
     pub open: Option<Arc<InMemoryLayer>>,
 }
 
+/// A result of an inmemory layer data being written to disk.
+pub struct LayersOnDisk {
```

Comment:
The reason for this to appear is the need to have a clear separation between layers when decide what to upload.
When the whole relish storage sync loop starts, we need to scan both local and remote storages and compare the layers, to see which files are missing.
Part of that is https://github.com/zenithdb/zenith/pull/686/files#diff-8ef63341d08bdfccc129283af160e1b6032defcd0bc929974510710cdf3d6478R217

`Layer` trait does not have much info to use when doing that, opposed to `DeltaLayer` and `ImageLayer` structs ergo the change.

------

Also, this way we've got solid Rust code, no extra vtables and other sorcery happens now which is way better IMO.

Reply:
This seems a bit fragile, or a leaky abstraction if you will. Currently, 'write_to_disk' returns zero/one image layer, and possibly some delta layers, but I consider that an implementation detail. It should be free to create any number of image and delta layers, if it thinks that it's useful.

You could call `filename()` and compare those. The file name is ultimately what we use to match up the files in S3 and on local disk. Or if dealing with filenames seems inconvenient, we could invent a new `LayerId` struct that contains all the same information that's currently in the filename.

> Also, this way we've got solid Rust code, no extra vtables and other sorcery happens now which is way better IMO.

Better how? Performance?

Reply:
> Better how? Performance?

Yes, that + it feels harder to use various existential types, as they're not accepted everywhere, you cannot have multiple traits defined in a generic way, etc., etc.

We're lucky to not to use such things on the crate boundaries much, otherwise some subtle bugs as https://github.com/rust-lang/rust/issues/88438 may occur for us too.

Reply:
> It should be free to create any number of image and delta layers, if it thinks that it's useful.

Sure thing, I've adjusted that to be a collection instead of Option now. I remember it was so before, but later some refactoring had changed it to be a single element and that confused me a bit.

> You could call filename() and compare those. The file name is ultimately what we use to match up the files in S3 and on local disk. Or if dealing with filenames seems inconvenient, we could invent a new LayerId struct that contains all the same information that's currently in the filename.

I like the idea and tried to use `PathBuf` everywhere and that almost worked. The biggest issue was the need to deal with LSNs for the remote timelines.
When we start, we have to read all timelines and
* determine the timeline with latest LSN to download
* react somehow on attempts to upload lower LSN layers

What are the ways to get the `disk_consistent_lsn` for a given remote timeline?
* read the metadata file
Doing that for every timeline is costly

* parse the file paths and get layer names with LSN data in them
That's what I do now, since I have to parse the paths anyway to extract `ZTenantId` and `ZTimelineId`anyway
Not sure if it's correct to derive the `disk_consistent_lsn` from this data, but simple to determine which timeline is the latest and download it, if needed.

* keep a special metadata file with latest `disk_consistent_lsn` and update it after every operation
Stas had proposed that, feels a bit scary IMO due to various concurrency reasons: we already fiddle with `metadata` file somewhat specially, by ensuring it's downloaded/uploaded as last + we still have to parse file paths to get the ids + that will bloat the code more

* something else??

So far I lean to operating the `LayerFileName` (https://github.com/zenithdb/zenith/pull/688/files#r720209536 feels like a very nice change to support that and in general) and lean to try Stas' proposal, but not in this PR series and rather when dealing with https://github.com/zenithdb/zenith/pull/689#issuecomment-931216193

Reply:
>     * read the metadata file
>       Doing that for every timeline is costly

Is it? Metadata files should be very small and reading them once at pageserver start is not a big deal.

Moreover, any other way to get the `disk_consistent_lsn` seems to be incorrect,
or maybe I don't really understand the problem you're trying to solve here.

Reply:
The problem can be vaguely formulated as "understand which layers we need to sync and how (download or upload)".
That has to be done from two directions:

* parse the local files and propagate the knowledge about them. Currently that's what pageserver does on init: now it unconditionally schedules the uploads of every startup-loaded timeline that has a proper metadata + pageserver uploads the frozen layers, covering all new files appearing
 
* parse the remote files, download some and keep the remote files with metadata in memory to compare with future uploads later

So, the main question is how to store the data in-memory to compare it later, and Heikki here proposed (if I've got him right) to directly store the `PathBuf` for every file, calculating the local path for remote files.
I almost like the idea, but the main culprit for me is the fact that I have to parse the remote file names anyway, to get the `ZTenantId` and `ZTimelineId`, so might make sense to parse the filenames also and ensure that the files are "correct".
Plus, I have to know some Lsn at least and I've used the one from file names to determine which timeline to download.

But the more I talk about the remote metadata files, the more I feel I have to parse it for every timeline and tenant after all unless we come up with something cleverer.